### PR TITLE
Improve likely class handling for pgo

### PR DESCRIPTION
**Pull request by davidwrighton**
- Move processing of likely class histogram into the JIT
- Fix cases where the devirtualization logic in crossgen2 behaved incorrectly in the presence of likely class histogram data
- Pre-process histogram at crossgen2 time to reduce the size of the pgo data
  - This removes 3/4 of the data from System.Private.CoreLib

Areas where I would particularly like to see comments
- The process by which crossgen2 calls into the JIT for preprocessing the type histogram. Currently its done via a free function exported from the jit dll, which isn't particularly clean. I'd like ideas on how to clean that up, or an agreement that its... ok.

### CODE REVIEW

**Comment by davidwrighton** [comment]
@dotnet/crossgen-contrib  @dotnet/jit-contrib

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 1]
```diff
@@ -147,6 +150,94 @@ public TextWriter Log
...
+            if (!hasTypeHistogram)
+            {
+                foreach (var elem in pgoData)
+                {
+                    yield return elem;
+                }
+            }
+            else
+            {
+                int currentObjectIndex = 0x1000000;
```
Comment (mangod9) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:174]: should this be defined as a constant?

**Comment by mangod9** [comment]
LGTM from cg2 side.

**Review by AndyAyersMS — COMMENTED** [review]
I'm ok with this approach of calling back into the jit. cc @BruceForstall in case he has any thoughts.

Seems like the histogram processing should be in some new file.

Do you know if this works properly for cross-bitness jitting? I am not sure how some of these types vary with target -- seems like `INT_PTR` might be tied to the host?

File: src/coreclr/inc/corjit.h [hunk 2]
```diff
@@ -404,24 +414,6 @@ class ICorJitInfo : public ICorDynamicInfo
             uint8_t **                pInstrumentationData         // OUT: `*pInstrumentationData` is set to the address of the instrumentation data.
             ) = 0;
 
-    // Get the likely implementing class for a virtual call or interface call made by ftnHnd
```
Comment (briansull) [inline src/coreclr/inc/corjit.h:407]: Changing the Jit Interface requires an update to JITEEVersionIdentifier located in jiteeversionguid.h

Comment (davidwrighton) [reply src/coreclr/inc/corjit.h:407]: Oops, I thought I had done that.

**Comment by davidwrighton** [comment]
@AndyAyersMS 
> Do you know if this works properly for cross-bitness jitting? I am not sure how some of these types vary with target -- seems like `INT_PTR` might be tied to the host?

INT_PTR is tied to the host, but that's ok for crossbitness jitting as the INT_PTR is simply the contract for JIT/VM communication for class types. For storage into the PE file there is a different format which is token based. (and sadly somewhat verbose, which is why the compressed representation here is quite important.)

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs (hunk 1 above)
Comment (davidwrighton) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:174]: Eh, the value is only used one place, and really just needs to be a large non-zero number. I'll leave a comment instead.

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 3]
```diff
@@ -147,6 +150,94 @@ public TextWriter Log
...
+            fixed(PgoInstrumentationSchema* pSchema = &nativeSchema[index])
+            {
+                fixed(byte* pInstrumentationData = &instrumentationData[0])
+                {
+                    IntPtr classType = getLikelyClass(pSchema, 2, pInstrumentationData, nativeSchema[index].ILOffset, out uint likelihood, out uint numberOfClasses);
+
+                    if (classType != IntPtr.Zero)
+                    {
+                        TypeDesc type = (TypeDesc)handleToObject[classType];
+                        if (compilationModuleGroup.VersionsWithType(type))
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:224]:
```suggestion
#if READYTORUN
                        if (compilationModuleGroup.VersionsWithType(type))
#endif
```

**Review by AndyAyersMS — APPROVED** [review]
We will likely want something like this for merging as well...

File: src/coreclr/jit/likelyclass.cpp [hunk 4]
```diff
@@ -0,0 +1,212 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+/*XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+XX                                                                           XX
+XX                           Likely Class Processing                         XX
+XX                                                                           XX
+XX   Parses Pgo data to find the most likely class in use at a given         XX
+XX   in a method. Used by both the JIT, and by crossgen                      XX
```
Comment (am11) [inline src/coreclr/jit/likelyclass.cpp:10]:
```suggestion
XX   offset in a method. Used by both the JIT, and by crossgen.             XX
```
maybe?

Comment (davidwrighton) [reply src/coreclr/jit/likelyclass.cpp]: Yes. I'm not sure where the IL offset bit of my comment went... I certainly remember writing it. Thanks for noting this.

**Comment by davidwrighton** [comment]
This should mostly address #45286

**Comment by BruceForstall** [comment]
@davidwrighton I'm late to the party here...

> The process by which crossgen2 calls into the JIT for preprocessing the type histogram. Currently its done via a free function exported from the jit dll, which isn't particularly clean. I'd like ideas on how to clean that up, or an agreement that its... ok.

Why can't it be on the JIT-EE interface?

**Pull request merged**